File: Lucy/2025/2-W02/2095_Delete_the_Middle_Node_of_a_Linked_List.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function deleteMiddle(head: ListNode | null): ListNode | null {
  let current: ListNode | null = head;
  let n = 1;
  while (current.next) {
    n++;
    current = current.next;
  }

  const middleNodeIndex: number = Math.floor(n / 2) - 1;
  if (middleNodeIndex === -1) return null;

  current = head;
  n = 0;
  while (current.next) {
    if (n === middleNodeIndex) {
      current.next = current.next.next;
      break;
    }

    current = current.next;
    n++;
  }

  return head;
}

File: Seori/LeetCode 75/Linked List/2095. Delete the Middle Node of a Linked List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def deleteMiddle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        
        # [A] Linked List의 총 길이 n을 찾아주는 함수
        def find_last_node(head: Optional[ListNode]) -> int:
            cnt = 1
            while head.next:
                head = head.next
                cnt += 1
            
            return cnt

        # [B] n번째 노드를 삭제해주는 함수
        def deleteNode(head: Optional[ListNode], n: int) -> None:
            for i in range(n-1):
                head = head.next
            head.next = head.next.next

        # 위 두 함수를 순서대로 실행해서, n//2번째 노드를 삭제한다.
        n = find_last_node(head)
        if n == 1: return
        deleteNode(head, n//2)

        return head